### DISCUSSION TITLE
Streaming Function Filters?

### DISCUSSION THREAD

**Discussion by Tyler-R-Kendrick** [microsoft/semantic-kernel#7006] [Q&A] [answered]
I'm trying to emulate some of the behavior in the python guidance library around constraining llm outputs, but for the dotnet semantic kernel sdk.

I'd like to create a pluggable component that is idiomatic with the design of SK, and not shim behavior around it; but, I'm struggling to find the appropriate points of extension.

Ideally, I'd have a StreamingFunctionInvocationFilter that I could use to intercept and cancel the output stream from OpenAI, but I don't see anything that looks like this.

The closest approximation I can find is to do something like this:

```csharp


public record struct FunctionStreamInvocationContext(
    FunctionInvocationContext Context,
    int TotalTokenCount,
    string ResultAggregate,
    string CurrentToken
);
public class ConstraintFunctionFilter(
    Func<FunctionStreamInvocationContext, bool> streamPredicate)
    : IFunctionInvocationFilter
{
    public async Task OnFunctionInvocationAsync(
        FunctionInvocationContext context,
        Func<FunctionInvocationContext, Task> next)
    {
        CancellationTokenSource tokenSource = new();
        var token = tokenSource.Token;
        var tokenCount = 0;
        var resultAggregate = string.Empty;
        await next(context);
        var result = context.Result;
        switch(result.GetValue<object>())
        {
            case IAsyncEnumerable<string> stream:
                await foreach (var item in stream.WithCancellation(token))
                {
                    resultAggregate += item;
                    var shouldContinue = streamPredicate(new(context, ++tokenCount, resultAggregate, item));
                    if (!shouldContinue)
                    {
                        tokenSource.Cancel();
                        break;
                    }
                }
                context.Result = new(context.Function, resultAggregate);
                break;
        }
    }
}

```
The idea here is to grab the stream, cancel it if criteria are met, and override the stream with a different result so subsequent filters don't continue enumeration of the original OpenAI request.

The problem is, I don't seem to be able to grab the byte data for tokens using this code - and I don't seem to be able to cancel the actual request (the stream continues generating tokens, since the cancellation only appears to apply to my enumeration of the stream and not the stream itself).

Any tips for intercepting this correctly in an idiomatic way?

**Comment by dmytrostruk** [comment, marked as answer]
Hi @Tyler-R-Kendrick , I think your approach is correct, here is an example how to use filters in streaming scenario:
https://github.com/microsoft/semantic-kernel/blob/fe38a5f5d6aa0e6f6ffb85c3154db7568493be6a/dotnet/samples/Concepts/Filtering/FunctionInvocationFiltering.cs#L167-L187

But if you want to get byte data for tokens or cancel actual HTTP request, you can define your interception and cancellation logic by extending `HttpClient` with `DelegatingHandler`:
https://github.com/microsoft/semantic-kernel/blob/fe38a5f5d6aa0e6f6ffb85c3154db7568493be6a/dotnet/src/Plugins/Plugins.MsGraph/Connectors/Client/MsGraphClientLoggingHandler.cs#L52-L59

You can then pass your `HttpClient` to chat completion service:
https://github.com/microsoft/semantic-kernel/blob/fe38a5f5d6aa0e6f6ffb85c3154db7568493be6a/dotnet/samples/Concepts/ChatCompletion/OpenAI_FunctionCalling.cs#L99-L108

I think at some point we can add `CancellationToken` to filter context, so it will be possible to monitor a cancellation, but I'm not sure that filters will get access to byte data from HTTP request, and I don't think they should have this capability, since it's HTTP-related logic, and there are existing mechanisms how to intercept it, like with example above. Please let me know if this resolves your scenario. Thanks!

**Comment by Tyler-R-Kendrick** [comment]
Just so I'm sure I understand the approach here, I would likely capture the cancellation token from the client request and request cancellation in the function filter. Is that correct?

I think that should be a sufficient workaround in the interim, but it seems that the cancellation token really should be passed down through function invocation. Should I make a feature request and PR for this behavior? Or is there something already in the pipeline for it?

**Reply by dmytrostruk** [reply]
> I would likely capture the cancellation token from the client request and request cancellation in the function filter. Is that correct?

That's correct.

> it seems that the cancellation token really should be passed down through function invocation. Should I make a feature request and PR for this behavior? Or is there something already in the pipeline for it?

I'm currently working on removing experimental flags from Filters feature, so I can create a PR for cancellation token in the same scope of work. If you could create a feature request, that would be great. Thanks!

**Reply by Tyler-R-Kendrick** [reply]
I've opened a request [here](https://github.com/microsoft/semantic-kernel/issues/7027). I'll close out this discussion. Thanks for your quick replies!

**Reply by dmytrostruk** [reply]
> I would likely capture the cancellation token from the client request and request cancellation in the function filter. Is that correct?
That's correct.

@Tyler-R-Kendrick A little clarification here, cancellation token would allow to see if any cancellation happened above the stack so it will be possible to stop the operation inside the filter. It would also allow to pass the original cancellation token from function invocation in filter to subsequent async operations when needed. But cancellation token itself is not capable of cancelling the operation, it is a responsibility of `CancellationTokenSource`. But even today, if for some reason you want to stop function execution, you don't need a cancellation token, you can throw an exception with details, like in this example:
https://github.com/microsoft/semantic-kernel/blob/775994e6c8f6a3f9dff331a4c94f56b8929bb7ae/dotnet/samples/Concepts/Filtering/FunctionInvocationFiltering.cs#L196-L208
